sound/soc/sof/ipc4-loader.c

```diff
@@ -391,6 +391,13 @@ int sof_ipc4_query_fw_configuration(struct snd_sof_dev *sdev)
 				goto out;
 			}
 			break;
+		case SOF_IPC4_FW_CFG_FAST_CLOCK_FREQ_HZ:
+			trace_sof_ipc4_fw_config(sdev, "Fast clock Hz", *tuple->value);
+			ipc4_data->max_core_frequency = *tuple->value;
+			/* Assume 400MHz if the received value is 0 */
+			if (!ipc4_data->max_core_frequency)
+				ipc4_data->max_core_frequency = 400000000;
```

Comment:
we shouldn't do this. This is generic code, the max_core_frequency should be in the chip descriptors or something private to each vendor/SOC.

Reply:
So, update the sof_dev_desc to contain max_core_frequency?
or rather add a new optional callback to snd_sof_dsp_ops (`get_dsp_max_frequency()`) and use that?

Reply:
Yes, something coming from the platform.

I am however confused by all this, we do have cases for Intel where we lower the clock to use e.g. 120 MHz. In that case, what is the relevance of the max_core_frequency? We should not assume that the cores always run flat out at the max frequency.

Not that I think of if, why do we even need a concept of max frequency?

Reply:
@plbossart This is just used as a fallback in case no guidance is given at what frequency to run (either from fw manifest or from topology). This fallback is same as the behaviour has been in all past SOF releases, e.g. run at max speed of DSP and clock-gate. In IPC4, driver HAS to provide this information, otherwise DSP will run at the lowest possible clock and that typically will not be enough and you will get errors. In IPC3, driver did not have to provide this information, so we didn't have the problem. So in summary, if you care about the DSP speed, you MUST provide correct data via manifest or topology. Otherwise DSP will run at the fast clock always.

Reply:
@kv2019i I don't fully understand why the manifest or topology would tell us as which frequency to run. That'd be either an optimistic or worst-case value, in real life one would need to track usages and set the frequency as needed with some sort of governor.

It seems like we are in the Neanderthal age of frequency scaling...Dare I say Jurassic age?

Reply:
@plbossart Don't shoot the messenger :) That is how the IPC4 FW operates and if we don't provide values, IPC4 firmwares will run at 38.4Mhz in all scenarios.

Reply:
@kv2019i I have lost track of how IPC4 is supposed to work. In the early days the firmware was supposed to be minimal and stupid, and only apply the decisions made by the mighty host driver. Now it seems the firmware can decide at which frequency to run based on module configurations? 

I am not getting the full picture on who does what....

Reply:
@plbossart I'll loop in @abonislawski who is implementing this on FW side. 
I think that "early days" philosophy still stands. FW doesn't have any policy, but just acts on the data the mighty host driver sends. Host tells you need X MCPS to run pipeline A (when creating pipeline A), and FW will bump freq to meet that. If host sets up two instances of pipeline A, FW needs to bump to 2X and so forth.

Reply:
@kv2019i can you remind me how the host keeps track of the total X MCPS needed to run pipeline A? I've lost the plot, and I still don't get what the concept of pipeline plays here given that modules are be assigned to different cores. The pipeline was always notional for IPC4, modules are what matters.

Reply:
@plbossart, we have the CPC value per modules, when a module is initialized we send this CPC, the firmware  adds these values together and that gives the whole CPC need for the initialized modules.
When the module is removed, it's CPC is also removed from the sum.

With the CPC the kernel is not directly telling the firmware to pick a certain freq, it just gives the firmware the info that a certain component needs X cycles to run, sum up all the X cycles and you get the frequency need.

The fallback to platform max is really  a fallback to ensure that we will not underclock the DSP, it is better to run it full speed than have audio artifacts and errors due to starvation.